Make JavaScriptCore ARC-safe!

LumioseSil opened the pull request:
#### a21d7ebd6e2dba68bc9c22cd873e87d73e837e93
<pre>
Make JavaScriptCore ARC-safe!

Reviewed by NOBODY (OOPS!).

Whatever checkType was supposed to do, all it did was return m_ptr, which is useless,
and the only other constexpr that uses it was RetainPtr.
This also confuses ARC, which could not follow the logic for some reason.

So now JavaScriptCore can run on ARC.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/59fc071cfa9db8c439d1129f5c1dfbc631700820

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/1984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2013 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2074 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/2908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/2059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2089 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/2054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/2908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2001 "Failed to checkout and rebase branch from PR 11299") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1781 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/1780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2755 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/1771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/1762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/1668 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/1661 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/1820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/1809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/2924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/1895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/1827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/2054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/2042 "Built successfully") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/1753 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/1773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/468 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/1925 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/2087 "Built successfully") | 
| | | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/3/builds/582 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WTF/wtf/RetainPtr.h:
```diff
@@ -140,8 +140,6 @@ template<typename T> class RetainPtr {
     enum AdoptTag { Adopt };
     constexpr RetainPtr(PtrType ptr, AdoptTag) : m_ptr(toStorageType(ptr)) { }
 
-    static constexpr PtrType checkType(PtrType ptr) { return ptr; }
```
The `checkType` method has a very specific purpose--to check that two types are assignable.  It should not be deleted.  See 0446f40c263cd6098fe566a9839339bc3158d98a for more details.

It purposely compiles to nothing since it's only there to do a sanity check using the compiler.

darinadler replied:
That’s right. If you are solving a problem by removing `checkType`, we would like to know what that problem is.

ddkilzer commented on Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
```diff
@@ -12990,13 +12990,15 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = 1C9051430BA9E8A70081E9D0 /* JavaScriptCore.xcconfig */;
 			buildSettings = {
+				CLANG_ENABLE_OBJC_ARC = YES;
```
This is the wrong place to set this Xcode variable.

Instead it should be added to `Source/JavaScriptCore/Configurations/Base.xcconfig`.

ddkilzer commented on Source/JavaScriptCore/API/JSContext.mm:
```diff
@@ -97,8 +88,10 @@ - (void)dealloc
 {
     m_exception.clear();
     JSGlobalContextRelease(m_context);
+#if !__has_feature(objc_arc)
     [_exceptionHandler release];
     [super dealloc];
+#endif
```
Everywhere you write `#if !__has_feature(objc_arc) ... #endif` you should just delete the code.

darinadler replied:
I agree.

ddkilzer commented on Source/JavaScriptCore/API/JSContext.mm:
```diff
@@ -61,14 +61,6 @@ - (JSGlobalContextRef)JSGlobalContextRef
     return m_context;
 }
 
-- (void)ensureWrapperMap
-{
-    if (!toJS([self JSGlobalContextRef])->wrapperMap()) {
-        // The map will be retained by the GlobalObject in initialization.
-        [[[JSWrapperMap alloc] initWithGlobalContextRef:[self JSGlobalContextRef]] release];
-    }
-}
```
Why delete the `JSWrapperMap`?

ddkilzer commented on Source/JavaScriptCore/API/ObjCCallbackFunction.mm:
```diff
@@ -90,15 +90,15 @@ void set(NSInvocation *invocation, NSInteger argumentNumber, JSContext *context,
     void set(NSInvocation *invocation, NSInteger argumentNumber, JSContext *context, JSValueRef argument, JSValueRef*) final
     {
-        JSValue *value = [JSValue valueWithJSValueRef:argument inContext:context];
+        __unsafe_unretained JSValue *value = [JSValue valueWithJSValueRef:argument inContext:context];
```
Why use `__unsafe_unretained` in these changes to this file?  What problem does this solve?

darinadler replied:
I have the same question and concern.

ddkilzer reviewed (changes requested):
r- due to:
- Review comments below that need to be fixed.
- Some `testapi` tests are failing on the jsc bots for macOS (probably due to autoreleased JavaScriptCore API objects that are being deallocated in an unexpected order from the autoreleasePool).

FYI, there's already a WIP patch over here:

Bug [221107](https://bugs.webkit.org/show_bug.cgi?id=221107): JavaScriptCore should build with ARC enabled

That patch also mostly works, except for the additional autoreleased objects occurring after enabling ARC.  That's the next step to investigate—how these objects are being autoreleased under ARC now.  (I have some ideas, but haven't had time to get back to the patch.  I should have time soon.)

Please reach out on Slack to discuss this more.  My username is the same there as it is here.

darinadler commented on Source/JavaScriptCore/API/JSAPIWrapperObject.mm:
```diff
@@ -52,7 +52,7 @@
         return;
 
-    JSC::Heap::heap(wrapperObject)->releaseSoon(adoptNS(static_cast<id>(wrapperObject->wrappedObject())));
+    JSC::Heap::heap(wrapperObject)->releaseSoon(adoptNS((__bridge id)((wrapperObject->wrappedObject()))));
```
```suggestion
    JSC::Heap::heap(wrapperObject)->releaseSoon(adoptNS((__bridge id)wrapperObject->wrappedObject()));
```
Too many parentheses.

darinadler commented on Source/JavaScriptCore/API/JSContext.mm:
```diff
@@ -81,13 +73,12 @@ - (instancetype)initWithVirtualMachine:(JSVirtualMachine *)virtualMachine
 
     m_virtualMachine = virtualMachine;
-    m_context = JSGlobalContextCreateInGroup(getGroupFromVirtualMachine(virtualMachine), 0);
+    m_context = JSGlobalContextRetain(JSGlobalContextCreateInGroup(getGroupFromVirtualMachine(virtualMachine), 0));
```
This looks wrong. A function with the word create in its name creates an object that is already retained. Retaining it a second time will cause a storage leak, unless we somehow remember to release it twice.

darinadler commented on Source/JavaScriptCore/API/JSWrapperMap.mm:
```diff
@@ -512,9 +514,9 @@ - (ConstructorPrototypePair)allocateConstructorAndPrototypeInContext:(JSContext
-        forEachProtocolImplementingProtocol(m_class, exportProtocol, ^(Protocol *protocol, bool&){
-            copyPrototypeProperties(context, m_class, protocol, prototype);
-            copyMethodsToObject(context, m_class, protocol, NO, constructor);
+        forEachProtocolImplementingProtocol(self->m_class, exportProtocol, ^(Protocol *protocol, bool &) {
+            copyPrototypeProperties(context, self->m_class, protocol, prototype);
+            copyMethodsToObject(context, self->m_class, protocol, NO, constructor);
```
What motivates this change? The `self->` should not be needed. And adding a space in `bool &` is simply changing away from WebKit coding style without giving a reason.

darinadler commented on Source/JavaScriptCore/API/ObjCCallbackFunction.mm:
```diff
@@ -727,13 +727,13 @@ JSObjectRef objCCallbackFunctionForMethod(JSContext *context, Class cls, Protoco
+    if (isInstanceMethod)
+        return objCCallbackFunctionForInvocation(context, invocation, CallbackInstanceMethod, cls, _protocol_getMethodTypeEncoding(protocol, sel, YES, YES));
+
+    // We need to copy the target Class because m_invocation doesn't retain it by default (and we don't want it to).
+    [invocation setTarget:[cls copy]];
```
This comment no longer makes when you replace the word "retain" with "copy".

What releases this copy we made of the class?

darinadler reviewed (changes requested):
I’m really glad you are tackling this. Much of it is correct, but quite a bit seems wrong.

LumioseSil committed:
Make JavaScriptCore ARC-safe!

Reviewed by NOBODY (OOPS!).

Whatever checkType was supposed to do, all it did was return m_ptr, which is useless,
and the only other constexpr that uses it was RetainPtr.
This also confuses ARC, which could not follow the logic for some reason.

So now JavaScriptCore can run on ARC.

LumioseSil committed:
Update Source/JavaScriptCore/API/JSAPIWrapperObject.mm

Co-authored-by: Darin Adler <33187212+darinadler@users.noreply.github.com>
